File: electron/services/workflowEngine.ts
import path from "node:path";
import fs from "node:fs";
import { PDFDocument } from "pdf-lib";
import type {
  DocumentInfo,
  DuplexComputation,
  LearningAnswers,
  PageParity,
  PageOrder,
  PassPlan,
  PrinterInfo,
  PrinterProfile,
  WorkflowSummary
} from "../../src/shared/types";

function pageRange(pageCount: number, parity: "odd" | "even"): number[] {
  const pages: number[] = [];
  for (let page = 1; page <= pageCount; page += 1) {
    if ((parity === "odd" && page % 2 === 1) || (parity === "even" && page % 2 === 0)) {
      pages.push(page);
    }
  }
  return pages;
}

function plan(
  pass: 1 | 2,
  description: string,
  pages: number[],
  order: PageOrder,
  parity: PageParity,
  insertsBlankTrailingSheet = false
): PassPlan {
  return {
    pass,
    description,
    pages: order === "reverse" ? [...pages].reverse() : pages,
    order,
    parity,
    insertsBlankTrailingSheet
  };
}

function resolveWorkflow(profile?: PrinterProfile): {
  firstPassParity: PageParity;
  secondPassParity: PageParity;
  firstPassOrder: PageOrder;
  secondPassOrder: PageOrder;
} {
  if (!profile) {
    return {
      firstPassParity: "odd",
      secondPassParity: "even",
      firstPassOrder: "normal",
      secondPassOrder: "normal"
    };
  }

  if (!profile.firstPassParity || !profile.secondPassParity || !profile.firstPassOrder) {
    if (profile.reloadMethod === "same_stack") {
      return {
        firstPassParity: "even",
        secondPassParity: "odd",
        firstPassOrder: "normal",
        secondPassOrder: "normal"
      };
    }

    return {
      firstPassParity: "odd",
      secondPassParity: "even",
      firstPassOrder: "reverse",
      secondPassOrder: "normal"
    };
  }

  return {
    firstPassParity: profile.firstPassParity,
    secondPassParity: profile.secondPassParity,
    firstPassOrder: profile.firstPassOrder,
    secondPassOrder: profile.secondPassParity === "even" ? profile.evenPagesOrder : "normal"
  };
}

export class WorkflowEngine {
  async createDocument(filePath: string): Promise<DocumentInfo> {
    const basename = path.basename(filePath);
    const bytes = fs.readFileSync(filePath);
    const pdf = await PDFDocument.load(bytes);

    return {
      path: filePath,
      fileName: basename,
      pageCount: pdf.getPageCount()
    };
  }

  createWorkflow(document: DocumentInfo, printer: PrinterInfo, profile?: PrinterProfile): WorkflowSummary {
    const oddPages = pageRange(document.pageCount, "odd");
    const evenPages = pageRange(document.pageCount, "even");
    const { firstPassParity, secondPassParity, firstPassOrder, secondPassOrder } = resolveWorkflow(profile);
    const firstPassPages = firstPassParity === "odd" ? oddPages : evenPages;
    const secondPassPages = secondPassParity === "odd" ? oddPages : evenPages;
    const insertsBlankTrailingSheet = firstPassPages.length < secondPassPages.length;

    return {
      document,
      printer,
      profile,
      firstPass: plan(
        1,
        `Print ${firstPassParity} pages first.`,
        firstPassPages,
        firstPassOrder,
        firstPassParity,
        insertsBlankTrailingSheet
      ),
      secondPass: plan(2, `Reload the paper stack, then print ${secondPassParity} pages.`, secondPassPages, secondPassOrder, secondPassParity),
      requiresLearning: !profile,
      finalStackGoal: "page_1_on_top"
    };
  }

  createCalibrationWorkflow(document: DocumentInfo, printer: PrinterInfo): WorkflowSummary {
    const evenPages = pageRange(document.pageCount, "even");

    return {
      document,
      printer,
      firstPass: plan(1, "Print calibration even pages first.", evenPages, "normal", "even"),
      secondPass: plan(2, "Calibration second pass is not used.", [], "normal", "odd"),
      requiresLearning: true,
      finalStackGoal: "page_1_on_top"
    };
  }

  createSecondPassPlan(document: DocumentInfo, answers: LearningAnswers): PassPlan {
    const evenPages = pageRange(document.pageCount, "even");
    const order = answers.reloadMethod === "same_stack" ? "reverse" : "normal";

    return plan(
      2,
      "Reload the printed stack exactly as it came out, then print the remaining pages.",
      evenPages,
      order,
      "even"
    );
  }

  learnFromCalibration(observedTopSheetPage: string): DuplexComputation {
    const parsed = Number.parseInt(observedTopSheetPage, 10);
    const firstPassParity: PageParity = Number.isNaN(parsed) || parsed <= 2 ? "even" : "odd";
    const secondPassParity: PageParity = firstPassParity === "even" ? "odd" : "even";
    const firstPassOrder: PageOrder = firstPassParity === "even" ? "normal" : "reverse";
    const evenPagesOrder: PageOrder = secondPassParity === "even" ? "normal" : "reverse";

    return {
      outputFace: parsed % 2 === 0 ? "down" : "up",
      firstPassParity,
      secondPassParity,
      firstPassOrder,
      evenPagesOrder,
      reloadMethod: firstPassParity === "even" ? "same_stack" : "flip_long_edge"
    };
  }

  learnFromAnswers(printer: PrinterInfo, answers: LearningAnswers): DuplexComputation {
    const parsed = Number.parseInt(answers.observedTopSheetPage, 10);
    const evenPagesOrder: PageOrder = answers.reloadMethod === "same_stack" ? "reverse" : "normal";
    const firstPassParity: PageParity = answers.reloadMethod === "same_stack" ? "even" : "odd";
    const secondPassParity: PageParity = firstPassParity === "even" ? "odd" : "even";
    const firstPassOrder: PageOrder = firstPassParity === "even" ? "normal" : "reverse";

    return {
      outputFace: parsed % 2 === 0 ? "down" : "up",
      firstPassParity,
      secondPassParity,
      firstPassOrder,
      evenPagesOrder,
      reloadMethod: answers.reloadMethod
    };
  }
}

File: cpp/core/workflow_engine.cpp
#include "workflow_engine.hpp"

#include <algorithm>
#include <charconv>
#include <string_view>

namespace duplexprint::core {
namespace {

std::vector<std::int32_t> page_range(const std::int32_t page_count, const PageParity parity) {
  std::vector<std::int32_t> pages;
  for (std::int32_t page = 1; page <= page_count; ++page) {
    const bool matches = (parity == PageParity::Odd && page % 2 == 1) ||
                         (parity == PageParity::Even && page % 2 == 0);
    if (matches) {
      pages.push_back(page);
    }
  }
  return pages;
}

PassPlan make_plan(
    const std::int32_t pass,
    std::string description,
    std::vector<std::int32_t> pages,
    const PageOrder order,
    const PageParity parity,
    const bool inserts_blank_trailing_sheet = false) {
  if (order == PageOrder::Reverse) {
    std::reverse(pages.begin(), pages.end());
  }

  return PassPlan {
      .pass = pass,
      .description = std::move(description),
      .pages = std::move(pages),
      .order = order,
      .parity = parity,
      .inserts_blank_trailing_sheet = inserts_blank_trailing_sheet,
  };
}

struct WorkflowResolution {
  PageParity first_pass_parity {PageParity::Odd};
  PageParity second_pass_parity {PageParity::Even};
  PageOrder first_pass_order {PageOrder::Normal};
  PageOrder second_pass_order {PageOrder::Normal};
};

WorkflowResolution resolve_workflow(const std::optional<PrinterProfile>& profile) {
  if (!profile.has_value()) {
    return {};
  }

  if (profile->reload_method == ReloadMethod::SameStack) {
    return WorkflowResolution {
        .first_pass_parity = PageParity::Even,
        .second_pass_parity = PageParity::Odd,
        .first_pass_order = PageOrder::Normal,
        .second_pass_order = PageOrder::Normal,
    };
  }

  return WorkflowResolution {
      .first_pass_parity = profile->first_pass_parity,
      .second_pass_parity = profile->second_pass_parity,
      .first_pass_order = profile->first_pass_order,
      .second_pass_order = profile->second_pass_parity == PageParity::Even ? profile->even_pages_order : PageOrder::Normal,
  };
}

int parse_int(const std::string_view value, const int fallback = 0) {
  int parsed = fallback;
  const auto* begin = value.data();
  const auto* end = value.data() + value.size();
  const auto result = std::from_chars(begin, end, parsed);
  if (result.ec != std::errc() || result.ptr != end) {
    return fallback;
  }
  return parsed;
}

OutputFace resolve_output_face(const int observed_top_sheet_page) {
  return observed_top_sheet_page != 0 && observed_top_sheet_page % 2 == 0 ? OutputFace::Down : OutputFace::Up;
}

}  // namespace

WorkflowSummary WorkflowEngine::create_workflow(
    const DocumentInfo& document,
    const PrinterInfo& printer,
    const std::optional<PrinterProfile>& profile) const {
  const auto odd_pages = page_range(document.page_count, PageParity::Odd);
  const auto even_pages = page_range(document.page_count, PageParity::Even);
  const auto workflow = resolve_workflow(profile);
  const auto& first_pass_pages = workflow.first_pass_parity == PageParity::Odd ? odd_pages : even_pages;
  const auto& second_pass_pages = workflow.second_pass_parity == PageParity::Odd ? odd_pages : even_pages;
  const bool inserts_blank_trailing_sheet = first_pass_pages.size() < second_pass_pages.size();

  return WorkflowSummary {
      .document = document,
      .printer = printer,
      .profile = profile,
      .first_pass = make_plan(
          1,
          "Print first pass.",
          first_pass_pages,
          workflow.first_pass_order,
          workflow.first_pass_parity,
          inserts_blank_trailing_sheet),
      .second_pass = make_plan(
          2,
          "Reload the paper stack, then print the remaining pages.",
          second_pass_pages,
          workflow.second_pass_order,
          workflow.second_pass_parity),
      .requires_learning = !profile.has_value(),
      .final_stack_goal = "page_1_on_top",
  };
}

WorkflowSummary WorkflowEngine::create_calibration_workflow(
    const DocumentInfo& document,
    const PrinterInfo& printer) const {
  return WorkflowSummary {
      .document = document,
      .printer = printer,
      .profile = std::nullopt,
      .first_pass = make_plan(1, "Print calibration even pages first.", page_range(document.page_count, PageParity::Even), PageOrder::Normal, PageParity::Even),
      .second_pass = make_plan(2, "Calibration second pass is not used.", {}, PageOrder::Normal, PageParity::Odd),
      .requires_learning = true,
      .final_stack_goal = "page_1_on_top",
  };
}

PassPlan WorkflowEngine::create_second_pass_plan(
    const DocumentInfo& document,
    const LearningAnswers& answers) const {
  const auto order = answers.reload_method == ReloadMethod::SameStack ? PageOrder::Reverse : PageOrder::Normal;
  return make_plan(
      2,
      "Reload the printed stack exactly as it came out, then print the remaining pages.",
      page_range(document.page_count, PageParity::Even),
      order,
      PageParity::Even);
}

DuplexComputation WorkflowEngine::learn_from_calibration(
    const std::string& observed_top_sheet_page) const {
  const int parsed = parse_int(observed_top_sheet_page, 0);
  const PageParity first_pass_parity = parsed <= 2 ? PageParity::Even : PageParity::Odd;
  const PageParity second_pass_parity = first_pass_parity == PageParity::Even ? PageParity::Odd : PageParity::Even;
  const PageOrder first_pass_order = first_pass_parity == PageParity::Even ? PageOrder::Normal : PageOrder::Reverse;
  const PageOrder even_pages_order = second_pass_parity == PageParity::Even ? PageOrder::Normal : PageOrder::Reverse;

  return DuplexComputation {
      .output_face = resolve_output_face(parsed),
      .first_pass_parity = first_pass_parity,
      .second_pass_parity = second_pass_parity,
      .first_pass_order = first_pass_order,
      .even_pages_order = even_pages_order,
      .reload_method = first_pass_parity == PageParity::Even ? ReloadMethod::SameStack : ReloadMethod::FlipLongEdge,
  };
}

DuplexComputation WorkflowEngine::learn_from_answers(const LearningAnswers& answers) const {
  const int parsed = parse_int(answers.observed_top_sheet_page, 1);
  const PageOrder even_pages_order = answers.reload_method == ReloadMethod::SameStack ? PageOrder::Reverse : PageOrder::Normal;
  const PageParity first_pass_parity = answers.reload_method == ReloadMethod::SameStack ? PageParity::Even : PageParity::Odd;
  const PageParity second_pass_parity = first_pass_parity == PageParity::Even ? PageParity::Odd : PageParity::Even;
  const PageOrder first_pass_order = first_pass_parity == PageParity::Even ? PageOrder::Normal : PageOrder::Reverse;

  return DuplexComputation {
      .output_face = resolve_output_face(parsed),
      .first_pass_parity = first_pass_parity,
      .second_pass_parity = second_pass_parity,
      .first_pass_order = first_pass_order,
      .even_pages_order = even_pages_order,
      .reload_method = answers.reload_method,
  };
}

}  // namespace duplexprint::core
